File: test_backend/src/main/java/com/myform/test/model/Sector.java
package com.myform.test.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import lombok.experimental.Accessors;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@Accessors(chain = true)
@ToString
@Table(name = "sectors")
public class Sector {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @NotNull
    @NotEmpty
    private String name;

    @ToString.Exclude
    @OneToMany(fetch = FetchType.EAGER, mappedBy = "parentSectorName", cascade = CascadeType.ALL)
    private List<Sector> sectors = new ArrayList<>();

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "parent_id")
    @JsonIgnore
    private Sector parentSectorName;

}

File: test_frontend/src/components/Sector.js
import React from 'react';

const Sector = ({sectors, temp}) => {
    let beforeSpace = ""
    return (
        <>
            {sectors.map(data => {
                if (data.parentId != null) {
                    beforeSpace = temp + "\u00a0\u00a0\u00a0\u00a0"
                }
                return (
                    <>
                        <option value={data.id}>{`${beforeSpace}  ${data.name}`}</option>
                        {data.sectors && <Sector temp={beforeSpace} sectors={data.sectors}/>}
                    </>
                );
            })}
        </>
    );
}

export default Sector